### PR TITLE
Pythonic data

### PR DESCRIPTION
**Pull request by jbrockmendel**
`MNLogit` is a priority.  `MNLogit` _will_ require getting into some of the weeds of `base.data`.  We're going to have to clean up `base.data` in the process.  Doing the cleanup before trying to implement new features will make life better, lead to smaller more focused diffs.

I claim this PR changes zero logic.  Review can be delegated; no subject-matter expertise is necessary.

### CODE REVIEW

File: statsmodels/base/data.py [hunk 1]
```diff
@@ -190,47 +191,43 @@ def handle_missing(cls, endog, exog, missing, **kwargs):
...
-            combined_names = ['endog', 'exog']
+            combined.extend([endog, exog])
+            combined_names.extend(['endog', 'exog'])
         else:
-            combined = (endog,)
-            combined_names = ['endog']
-            none_array_names += ['exog']
+            combined.append(endog)
+            combined_names.append('endog')
+            none_array_names.append('exog')
```
Comment (jbrockmendel) [inline statsmodels/base/data.py]: This can be further simplified, but I'm saving that for a follow-up to keep the diff easily-checked.

**Comment by josef-pkt** [comment]
Looks overall fine based on a quick read, but it is not needed for any MNLogit fixes, AFAIU
Some of the dropped ifs like `if combined_2d:` might be either needed or be more robust in case combined_2d or similar is None  (defensive programming)

some of the other ones are taste questions
I like dict(zip(...) better than comprehension, even if Guido or some python guys prefer "non-functional"

I don't see much difference in .append and += in terms of "pythonicity", but I also prefer append to add to an existing list.

I need to go over some details again, but based on the quick read most of it can be merged

**Comment by jbrockmendel** [comment]
>  but it is not needed for any MNLogit fixes

The flake8 fixes are not _necessary_ but are nice to separate from logic-affecting PRs to keep diffs focused.  (I think you commented to that effect on something I reviewed recently.)

The `handle_missing ` changes are only needed inasmuch as I won't be able to focus on other more directly-pertinent things while my eye is twitching with these just a short scroll away.

> I like dict(zip(...) better than comprehension

That's reasonable, and I avoided changing e.g. `combined = dict(zip(combined_names, lmap(drop_nans, combined)))` despite it being not-idiomatic.  It's really the

```
combined += (np.asarray(value_array),)
combined_names += [key]
[...]
combined = dict(zip(combined_names, combined))
```

that is unnecessarily obtuse.

**Comment by josef-pkt** [comment]
However, you cannot refactor statsmodels to your taste for fixing bugs, especially if it introduces or potentially introduces a lot of new bugs
One travis environment that I checked seems to have 7 new failures or errors, which means this is not an innocent stylistic cleanup.

(and I'm **very** risk adverse in changing core code)

**Comment by jbrockmendel** [comment]
> One travis environment that I checked seems to have 7 new failures or errors, which means this is not an innocent stylistic cleanup.

My case is admittedly weakened by the fact that I didn't bother to run the tests locally before pushing.  Because I expect pushback (and unrelated test failures #3991, #3956, #3796), I consider this kind of PR expository.

> However, you cannot refactor statsmodels to your taste for fixing bugs

Fundamentally the issue we've been going back and forth on for days is about willingness to delegate given the reality that there is more to do than one person can plausibly do.  But this is your house and you get to make the rules.

> (and I'm very risk adverse in changing core code)

You've mentioned that much of the "core" code is unfamiliar, written by ancient elders in far off monasteries.  And that the tests are not robust enough to be confident in them catching bugs if changes are made.  And I've argued that this is not a situation that a) is acceptable or b) gets better on its own.  Which is why we're using teamwork to chip away at the problem. By...

Having less duplication (#3796).  Having more reliable tests (#3893).  Having less ancient cruft (#3886, #3714, #3866, #3442).  Having cleaner Separation of Concerns (#3992).  Using code-quality tools (#3792).  Having consistent/predictable attributes/names (#3890, #3787, #3708, #3689, #3671)

None of those example are about "core" code, but they are all about taking a big problem and breaking it more manageable pieces.  Using teamwork.

**Comment by josef-pkt** [comment]
I think the big difference is in priorities, bug-fixes and enhancements versus cleanup.
And extremely spoken, I don't care about code cleanup at the current stage of statsmodels development. (Largely but not completely true.)

Our core is solid, in the sense that currently it works as "expected" with known bugs and problems. The "core of the core" is well tested, the computation and the most common use cases in terms of possible input types and options.
Where our core is fragile is on the fringes, parts, options, possible cases that currently don't have test coverage, but either work or if not then have most likely an open issue.

Refactoring and some enhancement in the core are fragile in that they change how this fringe without test coverage behaves. In those cases the refactoring or changes have to be useful "enough".
For example, what seemed to be a simple enhancement, adding pandas returns to `predict` turned out not to be simple at all. We have three or four rounds of bug fixes for cases that we didn't think about and didn't check or test.

My risk aversion comes from the point that I don't want to have several rounds of bugfixes after some refactorings or enhancements in the core, unless there is a large benefit of the changes. "Clean" code is not one of those benefits that is large enough, unless it really looks innocent.

It took scipy 10 years of cleanup and code consolidation after the early expansion oriented phase. But we are not there yet, we are still in the construction stage. The core IS getting consolidated, but every once in a while we need changes to account for new types of models and functionality. Those new types of models initially often exploit ambiguities in the fringes of the core or python's flexibility for "hacks", until those parts get settled again.
i.e. we need **demand driven** refactoring.

Aside: triaging issues without labels increased the bug count again.

Nevertheless, I still think this PR will be mergeable.

**Comment by jbrockmendel** [comment]
We can kick the horse if we want to be _really_ sure that its dead.  You've successfully talked me out of speculative "less bad is better" PRs going forward.

**Comment by josef-pkt** [comment]
I'm looking forward to your bugfix and enhancement PRs.

Note, I won't mind and it would be helpful to reopen and revisit some issues and PRs, when the demand arises and when I'm not supposed to be bugfixing and merging for a release (and when I'm not almost fully occupied with GSOC projects)

**Comment by jbrockmendel** [comment]
Is there a gameplan for fixing test_arima111_predict_exog_2127?

**Comment by josef-pkt** [comment]
test_arima111_predict_exog_2127 was fixed by #4004  (that should fix the problem not the symptom)

**Comment by jbrockmendel** [comment]
OK.  So should I re-push to get travis to re-run?

**Comment by josef-pkt** [comment]
I think you will have to rebase on master so that this branch picks up the changes and fixes

**Pull request closed without merging**